### ISSUE TITLE
Suggestion: disallow browser variables when SSR

### ISSUE THREAD

**Issue by lhz516** [yannickcr/eslint-plugin-react#2538]
For SSR apps, there's no DOM variables in node environment. Shall we have an eslint rule to disallow DOM variables in contructor? What about `no-dom-in-contructor`?
```js
constructor(props) {
  super(props)
  this.state = {}
  this.audio = document.createElement('audio') // Error when SSR, document is not defined
}
```
I'm intested in contributing this. Please let me know if it sounds good.

**Comment by ljharb** [comment]
If the purpose is to ensure SSR can work (which would be an excellent rule) there's a lot more things it would have to check:
 - in class/createReactClass components, no browser things in the constructor, render, or in any pre-mount lifecycle methods
 - in functional components, no browser things outside of the appropriate post-mount hooks

where "browser things" includes any non-language globals.

**Comment by lhz516** [comment]
@ljharb Yea, you are right. Probably `no-ssr-dom`/`no-dom-ssr` sounds better.

Eslint itself uses the package [`globals`](https://www.npmjs.com/package/globals) to get the global variables for different envs, we can just take these [browser ones](https://github.com/sindresorhus/globals/blob/master/globals.json#L227) and disable them in React SSR lifecycles and function components. 

Places need to disable browser things when SSR:
- Class components
  - constructor
  - componentWillMount (Deprecated)
  - UNSAFE_componentWillMount
  - render
  - static members
- Function components
  - All the function scope except post-mount hooks and inline functions that run in browser

For function components, it's tricky because users may have some custom inline functions (including hooks) that run at server/client/both sides. Sometimes the callback function is imported from another file. Eslint is not good at checking code logic to ensure which parts run in the browser. If we allow users to add ignores, users may end up adding a lot of ignores.

An easier way is to ignore all inner functions inside the function components. What do you think?

**Comment by ljharb** [comment]
Inner functions are going to be tricky with class components too. I'm not sure what's the best way to handle it.

**Comment by lucasflomuller** [comment]
Sounds like we already have an accepted rule for that here #2057

**Comment by ljharb** [comment]
Good call; closed as a duplicate of #2057.
